### PR TITLE
Remove specifics of downstream request queues from scheduler

### PR DESCRIPTION
**Pull request by whalebot-helmsman**
Implementing other types of downstream (only `FIFO/LIFO Disk/Memory` at this moment) queues without touching `Scheduler` code is a desirable outcome for this PR.

To achieve this, next things are done:
- `request_from_dict/request_to_dict` specifics moved to downstream queue code
- `crawler` is added as a parameter of constructor to disk queues to provide `spider` for `request_from_dict/request_to_dict` functions. `crawler` will be used to get other settings for downstream queues from `crawler.settings` 
- common constructor interface for every downstream queue `__init__(self, crawler, key)`
- `_newmq/_newdq` methods removed from `Scheduler` and `downstream_queue_cls` added to constructors of priority queues
- `priority` parameter removed from priority queues `push` method, so they decide internally what to use as priority
- support of `key` parameter in downstream queue's constructor require to rewrite code of `queuelib.pqueue.PriorityQueue` in `scrapy.pqueues.ScrapyPriorityQueue`, so we don't use `queuelib` version any more
- after several code refactorings there was no much code in `_SlotPriorityQueues` class. Remained code is merged to `DownloaderAwarePriorityQueue` class
- ability to create directories hierarchy is added to downstream disk queues
- as a side effect on disk queues for `DownloaderAwarePriorityQueue` is stored in hierarchy now `requests.queue/host/priority` instead `requests.queue/host-priority`. With a lot of host-priority combinations plain structure create a lot of files in a single directory. This negatively affects performance of creation/deletion downstream queues
- there is no need in `_Priority` structure any more, it is deleted

I decided to not unify constructor interface for priority queues and downstream queues.

After this PR something like https://github.com/scrapy/scrapy/issues/3723 will be much easier and we can provide implementations of downstream queues for different hosted and self-hosted message queue systems.

### CODE REVIEW

**Comment by whalebot-helmsman** [comment]
@Gallaecio 
> Any chance we can make these changes while keeping those signatures backward compatible?

Current signatures has a very short lifespan. Probability of someone using them is very low. May be you have some specific case in mind?

**Comment by Gallaecio** [comment]
> May be you have some specific case in mind?

No specific case, I’m just a bit worried that there might be someone affected.

**Comment by whalebot-helmsman** [comment]
Tests are failing for `py34` as we don't merged https://github.com/scrapy/scrapy/pull/3892

**Comment by whalebot-helmsman** [comment]
>  there might be someone affected.

`1.7` alive only for 19 days, and in worst case they can pin version

**Comment by whalebot-helmsman** [comment]
Restarted the build as Travis has connectivity issues this morning
> Identified - We have identified the production systems affected by this issue and have removed these from the pool. No new builds will be affected by this, existing builds in the pool will be allowed to finish but might have network connectivity issues. Please expect additional updates shortly. Thank you
Aug 8, 07:50 UTC

File: scrapy/pqueues.py [hunk 1]
```diff
@@ -29,91 +29,85 @@ def _path_safe(text):
...
 
-    def __len__(self):
-        return sum(len(x) for x in self.pqueues.values()) if self.pqueues else 0
+        self.curprio = min(startprios)
 
-    def __contains__(self, slot):
-        return slot in self.pqueues
+    def qfactory(self, key):
+        return self.downstream_queue_cls(self.crawler,
+                                         self.key + '/' + str(key))
```
[the remark this answers is not in the document]
Comment (whalebot-helmsman) [reply scrapy/pqueues.py]: done in https://github.com/scrapy/scrapy/pull/3884/commits/dd98567e44f72b48ae59e9f2808bf1e1886138b6

**Comment by Gallaecio** [comment]
I’m not sure if this is the right pull request to address this, but if we aim to eventually support message queues, should we refactor the scheduler internals to switch “disk” and “memory” by more abstract concepts (e.g. “persistent”, “volatile”)?

File: scrapy/squeues.py [hunk 2]
```diff
@@ -3,10 +3,28 @@
 """
 
 import marshal
+import os
+import os.path
```
Comment (Gallaecio) [inline scrapy/squeues.py]: :lipstick: The second import is not necessary.

**Comment by whalebot-helmsman** [comment]
> should we refactor the scheduler internals to switch “disk” and “memory” to more abstract concepts like “persistent” and “volatile”?

I think we should in future, not in this PR.

**Comment by whalebot-helmsman** [comment]
Thanks @Gallaecio , @kmike and @wRAR for review.

**Pull request merged**